core/impl/src/main/java/com/blazebit/persistence/impl/keyset/KeysetManager.java

```diff
@@ -57,15 +68,175 @@ public void initialize(List<OrderByExpression> orderByExpressions) {
         keysetLink.initialize(orderByExpressions);
     }
 
+    public void buildOptimizedKeysetPredicate(StringBuilder sb) {
+        KeysetLink keysetLink = getKeysetLink();
+        KeysetMode keysetMode = keysetLink.getKeysetMode();
+        Keyset keyset = keysetLink.getKeyset();
+        Serializable[] key = keyset.getTuple();
+        OrderByExpression extractedNonNullableExpression = null;
+
+        boolean hasNullableOrderBys = false;
+        for (OrderByExpression orderByExpression : orderByExpressions) {
+            if (orderByExpression.isNullable()) {
+                hasNullableOrderBys = true;
+                break;
+            }
+        }
+
+        extractedNonNullableExpression = orderByExpressions.get(0);
+
+        // We can only use row value based keyset predicates if the dbms supports row values and row value comparison and
+        // if all order bys are non-nullable because null elements would break the row value comparison.
+        if (hasNullableOrderBys || !dbmsDialect.supportsRowValueConstructor() || !dbmsDialect.supportsFullRowValueComparison()) {
+            // if the first order by is nullable, we cannot render an optimized form because we would need to include
+            // null handling involving disjunction on the top predicate level
+            if (extractedNonNullableExpression.isNullable() && (!extractedNonNullableExpression.isNullFirst() || key[0] == null)) {
```

Comment:
I still don't understand the condition on this line. If `key[0]` is != null, this would render the optimized form even if the expression is nullable, given that null first is configured. I don't understand what the null precedence and the actual value has to do with the actual rendering.